api/all/src/main/java/io/opentelemetry/api/internal/ConfigUtil.java

```diff
@@ -33,12 +35,17 @@ private ConfigUtil() {}
    */
   public static String getString(String key, String defaultValue) {
     String normalizedKey = normalizePropertyKey(key);
+
+    // Cloning in order to avoid ConcurrentModificationException
+    // see https://github.com/open-telemetry/opentelemetry-java/issues/6732
     String systemProperty =
-        System.getProperties().entrySet().stream()
-            .filter(entry -> normalizedKey.equals(normalizePropertyKey(entry.getKey().toString())))
-            .map(entry -> entry.getValue().toString())
-            .findFirst()
-            .orElse(null);
+        ((Properties) System.getProperties().clone())
```

Comment:
It would be nice to have a test recreating this CME and demonstrating that this fixes it. However, I spent a decent amount of time trying to hammer this with concurrent reads and writes and was unable to recreate it, so it could be specific to some version of java or the jdk.

I'm content give this a shot even without strong proof that it resolves it.

Reply:
Looks like it does not reproduces on jdk 11+ probably because of https://github.com/openjdk/jdk11u/blob/cee8535a9d3de8558b4b5028d68e397e508bef71/src/java.base/share/classes/java/util/Properties.java#L159-L165
On jdk8 it reproduces with
```java
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public class Main {
    public static void main(String[] args) {
        int threads = 4;
        AtomicInteger counter = new AtomicInteger();
        Executor executor = Executors.newFixedThreadPool(threads);
        for (int i = 0; i < threads; i++) {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    while (true) {
                        String property = "prop " + counter.getAndIncrement();
                        System.setProperty(property, "a");
                        System.getProperties().remove(property);
                    }
                }
            });
        }

        while (true) {
            System.getProperties().entrySet().stream()
                    .filter(
                            entry -> "x".equals(entry.getKey().toString())
                    )
                    .map(entry -> entry.getValue().toString())
                    .findFirst()
                    .orElse(null);
        }
    }
}
```
`ConcurrentModificationException` does not reproduce with the proposed fix. It doesn't reproduce with
```java
Properties properties = System.getProperties();
synchronized (properties) {
    properties.entrySet().stream()
            .filter(
                    entry -> "x".equals(entry.getKey().toString())
            )
            .map(entry -> entry.getValue().toString())
            .findFirst()
            .orElse(null);
}
```
either. Unlike the proposed fix this version does not make extra copies of system properties.

Reply:
Good find @laurit!

I know there was some skepticism about taking a lock on `System.getProperties()`, but it seems ok to me. The work we're doing inside the lock is narrow and its hard for me to imagine how we could end up in a dead lock scenario.